pkg/credentialprovider/plugin/config.go

```diff
@@ -29,24 +31,84 @@ import (
-// into the internal CredentialProviderConfig type.
-func readCredentialProviderConfigFile(configPath string) (*kubeletconfig.CredentialProviderConfig, error) {
+// readCredentialProviderConfig receives a path to a config file or directory.
+// If the path is a directory, it reads all "*.json" files in lexicographic order,
+// decodes them, and merges their entries into a single CredentialProviderConfig object.
+// If the path is a file, it decodes the file into a CredentialProviderConfig object directly
+func readCredentialProviderConfig(configPath string) (*kubeletconfig.CredentialProviderConfig, error) {
 	if configPath == "" {
 		return nil, fmt.Errorf("credential provider config path is empty")
 	}
 
-	data, err := os.ReadFile(configPath)
+	fileInfo, err := os.Stat(configPath)
 	if err != nil {
-		return nil, fmt.Errorf("unable to read external registry credential provider configuration from %q: %w", configPath, err)
+		return nil, fmt.Errorf("unable to access path %q: %w", configPath, err)
 	}
 
-	config, err := decode(data)
-	if err != nil {
-		return nil, fmt.Errorf("error decoding config %s: %w", configPath, err)
+	var configs []*kubeletconfig.CredentialProviderConfig
+	var configFiles []string
+
+	if fileInfo.IsDir() {
+		entries, err := os.ReadDir(configPath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read directory %q: %w", configPath, err)
+		}
+
+		// Filter and sort *.json/*.yaml/*.yml files in lexicographic order
+		for _, entry := range entries {
+			ext := filepath.Ext(entry.Name())
+			if !entry.IsDir() && (ext == ".json" || ext == ".yaml" || ext == ".yml") {
+				configFiles = append(configFiles, filepath.Join(configPath, entry.Name()))
+			}
+		}
+		sort.Strings(configFiles)
+
+		if len(configFiles) == 0 {
+			return nil, fmt.Errorf("no configuration files found in directory %q", configPath)
+		}
+	} else {
+		configFiles = append(configFiles, configPath)
+	}
+
+	for _, filePath := range configFiles {
+		data, err := os.ReadFile(filePath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read file %q: %w", filePath, err)
+		}
+
+		config, err := decode(data)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding config %q: %w", filePath, err)
+		}
+		configs = append(configs, config)
+	}
+
+	// Merge all configs into a single CredentialProviderConfig
+	// When there are duplicate provider names, the last one overrides previous ones
+	mergedConfig := &kubeletconfig.CredentialProviderConfig{}
+	providerMap := make(map[string]kubeletconfig.CredentialProvider)
+
+	// Track the order of providers as they appear in the files
+	var providerNames []string
+	seenProviders := sets.NewString()
+
+	for _, config := range configs {
+		for _, provider := range config.Providers {
+			providerMap[provider.Name] = provider
+			// Only add to the order list if we haven't seen this provider before
+			if !seenProviders.Has(provider.Name) {
```

Comment:
if multiple files define the same provider by name, this will silently drop the latter ones... I would think failing the unique name validation (which is what happens when we merge them all together and return) would be better

Reply:
gotcha. Will go back to that one!